src/linear_algebra/vectors.lean

```diff
@@ -0,0 +1,56 @@
+/-
+Copyright (c) 2021 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak, Kyle Miller
+-/
+
+import data.matrix.basic
+import data.matrix.notation
+import tactic.fin_cases
+
+/-!
+# Vectors
+
+This module provides lemmata to more easily manipulate explicit vectors in $R^2$ and $R^3$
+for $R$ a commutative ring.
+
+## Notation
+
+The locale `vectors` gives the following notation:
+
+* `⬝` for dot products
+
+## Tags
+
+vectors
+-/
+
+variable {R : Type*}
+
+lemma vec2_eq {a₀ a₁ b₀ b₁ : R} (h₀ : a₀ = b₀) (h₁ : a₁ = b₁) :
+  ![a₀, a₁] = ![b₀, b₁] :=
+by { ext x, fin_cases x; assumption }
+
+lemma vec3_eq {a₀ a₁ a₂ b₀ b₁ b₂ : R} (h₀ : a₀ = b₀) (h₁ : a₁ = b₁) (h₂ : a₂ = b₂) :
+  ![a₀, a₁, a₂] = ![b₀, b₁, b₂] :=
+by { ext x, fin_cases x; assumption }
+
+variable [comm_ring R]
+
+lemma vec2_add {a₀ a₁ b₀ b₁ : R} :
+  ![a₀, a₁] + ![b₀, b₁] = ![a₀ + b₀, a₁ + b₁] :=
+by { ext x, fin_cases x; refl }
+
+lemma vec3_add {a₀ a₁ a₂ b₀ b₁ b₂ : R} :
+  ![a₀, a₁, a₂] + ![b₀, b₁, b₂] = ![a₀ + b₀, a₁ + b₁, a₂ + b₂] :=
+by { ext x, fin_cases x; refl }
```

Comment:
These only need `has_add R`

Reply:
Unfortunately, I didn't know how to get rid of `variable [comm_ring α] ` in `data/matrix/notation` on line 239. Can you please help me?

Reply:
Without the variable, the proof of `cross_product` being multilinear fails on line 52 in `linear_algebra/cross_product` when caling `apply vec3_eq`.

Reply:
I found the problem. It will take me some time to solve it. It was not about `vec3_eq` at all.

Reply:
The problem was in `vec3_add` leaving a different state if the instance was only `[has_add α]`. The current (hopefully temporary) solution is:
```
variable [add_semigroup α]
/-- TODO: understand why `[has_add α]` is not enough for `pi.add_apply` correctly work
          in `linear_algebra/cross_product` on line 51 -/
```
Imma create MWE for the specific issue with associativity (happens also when using `rw` instead of `simp`) and post it to Zulip.

Reply:
Discussion:
https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/has_add.20versus.20add_semigroup

Reply:
Fixed using Kevin Buzzard's suggestion.